File: libs/ml/DEPRECATED_CoreCrypto/CoreACiphers.ml
(* -------------------------------------------------------------------- *)
open Bytes
open Evp.RSA

(* -------------------------------------------------------------------- *)
type modulus  = bytes
type exponent = bytes

(* -------------------------------------------------------------------- *)
type sk = RSASKey of CoreKeys.rsaskey
type pk = RSAPKey of CoreKeys.rsapkey

type plain = bytes
type ctxt  = bytes

let gen_key () = 
    let k  = Evp.RSA.genkey 2048 65537 in
    let pk = RSAPKey (abytes k.k_mod, abytes k.k_pub_exp) in
    match k.k_prv_exp with 
      | Some sk -> (RSASKey (abytes k.k_mod,abytes sk),pk)
      | None -> failwith "RSA_generate_key returned an empty private key"

(* -------------------------------------------------------------------- *)
let encrypt_pkcs1 (RSAPKey (m, e)) (x : plain) =
  let r = create () in
  setkey r { k_mod = cbytes m; k_pub_exp = cbytes e; k_prv_exp = None; };
  let y = abytes (encrypt r false PD_PKCS1 (cbytes x)) in
  fini r; y

(* -------------------------------------------------------------------- *)
let decrypt_pkcs1 (RSASKey (m, e)) (x : ctxt) =
  let r = create () in
  setkey r { k_mod = cbytes m; k_pub_exp = cbytes e; k_prv_exp = Some (cbytes e); };
  let y = abytes (decrypt r true PD_PKCS1 (cbytes x)) in
  fini r; Some y

File: libs/fs/CoreCrypto/CoreACiphers.fs
(* Copyright (C) 2012--2015 Microsoft Research and INRIA *)

module CoreACiphers
open Bytes

open Org.BouncyCastle.Math
open Org.BouncyCastle.Crypto
open Org.BouncyCastle.Security
open Org.BouncyCastle.Crypto.Generators
open Org.BouncyCastle.Crypto.Encodings
open Org.BouncyCastle.Crypto.Engines
open Org.BouncyCastle.Crypto.Parameters

type modulus  = bytes
type exponent = bytes

type sk = RSASKey of CoreKeys.rsaskey
type pk = RSAPKey of CoreKeys.rsapkey

type plain = bytes
type ctxt  = bytes

//needs to be peer-reviewed and pen-tested
let gen_key () : sk * pk =
    let kparams  = new RsaKeyGenerationParameters(BigInteger("65537"),new SecureRandom(), 2048, 80) in
    let kgen     = new RsaKeyPairGenerator() in
        kgen.Init(kparams);
        let kpair = kgen.GenerateKeyPair() in
        let pkey  = (kpair.Public  :?> RsaKeyParameters ) in
        let skey  = (kpair.Private :?> RsaPrivateCrtKeyParameters) in
            (RSASKey(abytes (skey.Modulus.ToByteArrayUnsigned()), abytes(skey.Exponent.ToByteArrayUnsigned())), 
             RSAPKey(abytes (pkey.Modulus.ToByteArrayUnsigned()), abytes(pkey.Exponent.ToByteArrayUnsigned()))
            )

let encrypt_pkcs1 (RSAPKey (m, e)) (plain : plain) =
    let m, e   = new BigInteger(1, cbytes m), 
                 new BigInteger(1, cbytes e) in
    let engine = new RsaEngine() in
    let engine = new Pkcs1Encoding(engine) in

    engine.Init(true, new RsaKeyParameters(false, m, e))
    abytes (engine.ProcessBlock(cbytes plain, 0, length plain))

let decrypt_pkcs1 (RSASKey (m, e)) (ctxt : ctxt) =
    let m, e   = new BigInteger(1, cbytes m), 
                 new BigInteger(1, cbytes e) in
    let engine = new RsaEngine() in
    let engine = new Pkcs1Encoding(engine) in

    try
        engine.Init(false, new RsaKeyParameters(true, m, e))
        Some (abytes (engine.ProcessBlock(cbytes ctxt, 0, length ctxt)))
    with :? InvalidCipherTextException ->
        None
